src/averageAges.js

```diff
@@ -55,10 +77,36 @@
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const mothers = people.filter(mother => people.some(person =>
+    person.mother === mother.name));
+
+  let kids = people.filter(mummy => mummy.mother != null);
+
+  if (onlyWithSon === true) {
+    kids = people.filter(son => son.mother != null && son.sex === 'm');
+  }
+
+  let ageSum = 0;
+  let numberOfPairs = 0;
+
+  kids.forEach(kid => {
+    const mother = mothers.find(mummy => mummy.name === kid.mother);
+
+    if (mother) {
+      const ageDifference = kid.born - mother.born;
+
+      ageSum += ageDifference;
+      numberOfPairs++;
+    }
+  });
+
+  if (numberOfPairs > 0) {
+    const averageAgeDiff = ageSum / numberOfPairs;
+
+    return parseFloat(averageAgeDiff.toFixed(2));
+  } else {
+    return 0; // Zabezpieczenie przed dzieleniem przez zero
```

Comment:
The check for 'numberOfPairs > 0' is valid to avoid division by zero. However, consider whether returning 0 is the best approach. It might be more informative to return `NaN` to indicate that there are no valid pairs for the calculation.